### PR TITLE
Issue25 add notification function to notify hods about edit of allocation

### LINKED ISSUE #25: #25add notification function to notify HoDs about edit of allocation

### PR DESCRIPTION
**Pull request by kylelhk**
## Change Summary 📝

1.	Added a conditional notification pop-up alert for HoDs in nav bar that appears whenever any comment_status in the workload_allocation table is “Unread” AND the dept_id of the commenter equals to that of the current HoD user

2.	Adjusted nav bar CSS to accommodate the notification pop-up

3.	Created a Comment History page which is only assessable by HoDs, functions included:
        (a)   Role-based access control - HoDs can only view the comments made by the staff from their respective department.
        (b)   Filtering of comments based on staff number, comment contents and comment status
        (c)   Download of comment history report in CSV format
        (d)   Redirecting HoDs to view the workload entry corresponding to a particular comment

---

## Change Form 📋
Please mark as NA if not applicable. If a certain criterion is not met, kindly provide a reason.

- [x] The pull request title has an issue number.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] Relevant tests (if applicable) have been added or updated.

---

## Any background context you want to provide? 🖼️

Self-tests conducted:

A set of test data has been added into app.db (with record in seed.py) for testing the new functions.

3 new users:
{'username': 12345678, 'password': 'password'},  # Staff (dept_id = 2)
{'username': 22222221, 'password': 'password'},  # HoD for Physics (dept_id = 1)
{'username': 22222223, 'password': 'password'},  # HoD for CSSE (dept_id = 3)
image

1 new workload assignment:
image

1.	Notification alert:
(a)	22222221 does not get the notification in nav bar as all the comments in the database are from staff in other departments.
(b)	22222222 and 22222223 get the notification.
2.	Comment History:
(a)	22222221 can’t see anything in the Comment History page.
(b)	22222222 only see the comment made by 12345678 in the Comment History page (both having the same dept_id “2”), and the corresponding comment_status in the database is changed from Unread to Read after the comments have been loaded and displayed to the user.  Subsequently, the notification no longer pops up and if the user visits the Commit History page again, the comment status of the comments previously read becomes Read.
(c)	22222223 only see the comment made by 33333333 in the Comment History page (both having the same dept_id “3”).  The rest of the test result is the same as that of the test for user 22222222.
3.	HoS, Staff and Admin can’t see the nav link for the Comment History page in the nav bar.
6.	The filtering and CSV export functions work as expected.
7.	If the View Workload page is accessed through the redirection function in the Comment History page, only the entry related to the corresponding comment in the Comment History will be shown.
8.	Sometimes, in the SQLite GUI, the comment_status may not be updated by refreshing the table.  In that case, we have to remove the database and add it again to see the changes.

All the test results above meet the desired outcomes.

Possible enhancement:

Set up Flask-SocketIO for the server to push notifications to the client in real-time without the need for polling (sending a GET request to the server every 1 second to check if there is any unread comment)

Other notes:

1.	The previously discussed nav bar CSS issue in the dashboard and assign pages is yet to be fixed by Wendy.
2.	Modifications were made to @app.route('/view_workload') and view_workload.html for implementing and testing the redirection function in the Comment History table (via the View Workload button).  Though the modifications are based on the latest commit in the branch for Issue 14, the assignee(s) of the View Workload page functions should be notified of these changes and any potential conflict should be discussed and addressed.

## Reviewers 🕵️

(Optional) Suggest reviewers that might be interested in this PR.

### CODE REVIEW

File: app/static/css/base.css
Comment (yylocky) [inline app/static/css/base.css:1]: Navbar CSS issue in the dashboard and assign pages will need to further addressed with @WendyW09 to fix.

**Review by yylocky — APPROVED** [review]
Great implementation, All points mentioned in "change summary" have been tested with new dataset and meet all the requirement 🥇 

😄 Something might be further improved:
When user stays on "comment history" page, If there is new unread comment, popup notification will not show, user will need to manually refresh the page to view new incoming notification. "comment status" fetching logic could be further improved to solve this problem. But overall the notification is doing a great job on all of the other pages 👍🏻

File: app/static/css/base.css
Comment (kylelhk) [reply app/static/css/base.css:1]: @WendyW09 As discussed previously on Teams, please help fix the nav bar CSS issue in the dashboard and assign pages (Issue 31).  Thanks.

**Comment by kylelhk** [comment]
> Great implementation, All points mentioned in "change summary" have been tested with new dataset and meet all the requirement 🥇
> 
> 😄 Something might be further improved: When user stays on "comment history" page, If there is new unread comment, popup notification will not show, user will need to manually refresh the page to view new incoming notification. "comment status" fetching logic could be further improved to solve this problem. But overall the notification is doing a great job on all of the other pages 👍🏻

Good spot!  This can be a function enhancement in Deliverable 3 via setting up Flask-SocketIO for push notification.  Thanks.

**Pull request merged**